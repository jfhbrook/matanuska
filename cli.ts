import { getTracer } from './debug';
import { Exit, ExitCode } from './exit';
import { RuntimeFault, UsageFault } from './faults';
import { ConsoleHost, Host, Level } from './host';

const tracer = getTracer('main');

/**
 * Command line arguments.
 */
export type Argv = typeof process.argv;

/**
 * Command line environment.
 */
export type Env = typeof process.env;

/**
 * Exit the command.
 */
export type ExitFn = typeof process.exit;

/**
 * Options for configuring the CLI.
 */
export interface ConfigOptions {
  /**
   * An optional logging level. Defaults to Info.
   */
  level?: Level;
}

/**
 * Options for running a CLI.
 */
export interface RuntimeOptions {
  /**
   * A Host instance. Defaults to ConsoleHost.
   */
  host?: Host;

  /**
   * A command to exit the host. Defaults to process.exit.
   */
  exit?: ExitFn;

  /**
   * Command line arguments.
   */
  argv: Argv;

  /**
   * Command line environment.
   */
  env: Env;
}

export type Run<C> = (config: C, host: Host) => Promise<void>;

/**
 * A CLI command.
 */
export interface Cli<C extends ConfigOptions> {
  /**
   * Parse command line arguments and environment into a config.
   *
   * @param argv Command line arguments.
   * @param env Command line environment.
   * @returns A config.
   */
  parseArgs(argv: Argv, env: Env): C;

  /**
   * Run the CLI command.
   *
   * @param config A config.
   * @param host A Host.
   * @returns A promise that resolves on completion.
   */
  run: Run<C>;
}

/**
 * A main function.
 *
 * @param options CLI options.
 * @returns A promise that resolves on completion.
 */
export type Main = (options: RuntimeOptions) => Promise<void>;

/**
 * Create a main function from a CLI command.
 *
 * @param cli A CLI command.
 * @returns A main function.
 */
export function cli<C extends ConfigOptions>(cli: Cli<C>): Main {
  let error: any = null;
  return async function main({
    host: overriddenHost,
    exit: overriddenExit,
    argv,
    env,
  }: RuntimeOptions): Promise<void> {
    const host: Host = overriddenHost || new ConsoleHost();
    const exit: ExitFn = overriddenExit || process.exit;

    function errorExit(error: any): void {
      exit(
        typeof error.exitCode === 'number' ? error.exitCode : ExitCode.Software,
      );
    }

    function errorHandler(error: any): void {
      reportError(error, host);
      errorExit(error);
    }

    await tracer.span('cli', async () => {
      //
      // Errors in the main coroutine should be correctly handled. However,
      // we also need to handle errors thrown by unmanaged EventEmitters and/or
      // Promises that aren't awaited in the main coroutine.
      //
      process.on('uncaughtException', errorHandler);
      process.on('unhandledRejection', errorHandler);

      try {
        const config = cli.parseArgs(argv, env);
        host.setLevel(config.level);
        await cli.run(config, host);
      } catch (err) {
        reportError(err, host);
        error = err;
      }

      process.removeListener('uncaughtException', errorHandler);
      process.removeListener('unhandledRejection', errorHandler);
    });

    if (error) {
      errorExit(error);
      return;
    }

    // For consistency, explicitly exit instead of allowing Node to run the
    // command to completion.
    exit(0);
  };
}

/**
 * Report an error.
 *
 * @param err The object to report as an error.
 * @param host A Host instance.
 */
export function reportError(err: any, host: Host): void {
  tracer.spanSync('reportError', () => {
    if (err instanceof UsageFault) {
      host.writeLine(err);
    }

    // Handle intentional exits.
    if (err instanceof Exit) {
      // TODO: Should the user be able to access this log in a release build?
      tracer.trace(`Exit ${err.exitCode}`);
      if (err.message.length) {
        host.writeInfo(err.message);
      }
      return;
    }

    if (err.format) {
      try {
        host.writeException(err);
      } catch (_) {
        const fault = RuntimeFault.fromError(err, null);
        host.writeException(fault);
      }
    } else {
      const fault = RuntimeFault.fromError(err, null);
      host.writeException(fault);
    }
  });
}
