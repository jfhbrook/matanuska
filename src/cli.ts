import { Config } from './config';
import { Exit } from './exit';
import { UsageFault } from './faults';
import { ConsoleHost, Host, LoggingOptions } from './host';

/**
 * Command line arguments.
 */
export type Argv = typeof process.argv;

/**
 * Command line environment.
 */
export type Env = typeof process.env;

/**
 * Exit the command.
 */
export type ExitFn = typeof process.exit;

/**
 * Options for running a CLI.
 */
export interface CliOptions {
  /**
   * A Host instance. Defaults to ConsoleHost.
   */
  host?: Host;

  /**
   * A command to exit the host. Defaults to process.exit.
   */
  exit?: ExitFn;

  /**
   * Command line arguments.
   */
  argv: Argv;

  /**
   * Command line environment.
   */
  env: Env;
}

export type Run<C> = (config: C, host: Host) => Promise<void>;

/**
 * A CLI command.
 */
export interface Cli<C extends LoggingOptions> {
  /**
   * Parse command line arguments and environment into a config.
   *
   * @param argv Command line arguments.
   * @param env Command line environment.
   * @returns A config.
   */
  parseArgs(argv: Argv, env: Env): C;

  /**
   * Run the CLI command.
   *
   * @param config A config.
   * @param host A Host.
   * @returns A promise that resolves on completion.
   */
  run: Run<C>;
}

/**
 * A main function.
 *
 * @param options CLI options.
 * @returns A promise that resolves on completion.
 */
export type Main = (options: CliOptions) => Promise<void>;

/**
 * Create a main function from a CLI command.
 *
 * @param cli A CLI command.
 * @returns A main function.
 */
export function cli<C extends LoggingOptions>(cli: Cli<C>): Main {
  return async function main({
    host: overriddenHost,
    exit: overriddenExit,
    argv,
    env,
  }: CliOptions): Promise<void> {
    const host: Host = overriddenHost || new ConsoleHost();
    const exit: ExitFn = overriddenExit || process.exit;

    try {
      const config = cli.parseArgs(argv, env);
      host.setLevel(config.level);
      // NOTE: It is up to the main function to initialize and close the
      // host.
      await cli.run(config, host);
    } catch (err) {
      // CLI usage faults are appropriate to log onto the console.
      if (err instanceof UsageFault) {
        console.log(err.message);
        exit(err.exitCode);
      }

      // Handle successful exits.
      if (err instanceof Exit) {
        if (err.message.length) {
          host.writeInfo(err.message);
        }
        exit(err.exitCode);
      }

      if (typeof err.exitCode === 'number') {
        host.writeException(err);
        exit(err.exitCode);
      }

      // TODO: Handle other Faults and Errors
      throw err;
    }

    // For consistency, explicitly exit instead of allowing Node to run the
    // command to completion.
    exit(0);
  };
}
