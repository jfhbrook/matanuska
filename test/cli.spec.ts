import { basename } from 'node:path';

import { describe, expect, test } from 'vitest';

import { run } from './helpers/cli';
import { EXAMPLES } from './helpers/files';

describe('examples', () => {
  for (const path of Object.keys(EXAMPLES)) {
    const name = basename(path);
    switch (name) {
      // TODO: Some scripts will need mocked input
      default:
        test(name, async () => {
          const { exitCode, host } = await run([path], process.env);
          expect({
            exitCode,
            stdout: host.outputStream.output,
            stderr: host.errorStream.output,
          }).toMatchSnapshot();
        });
    }
  }
});
