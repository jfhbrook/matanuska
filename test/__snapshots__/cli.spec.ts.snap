// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`examples > 001-hello-world.bas 1`] = `
{
  "exitCode": 0,
  "stderr": "",
  "stdout": "hello world
goodbye world
",
}
`;

exports[`examples > 002-invalid-line-number.bas 1`] = `
{
  "exitCode": 70,
  "stderr": "/home/josh/matanuska/examples/002-invalid-line-number.bas:<R3>:0 (after line 20): error: Expected line number
  "foo"
  ^
",
  "stdout": "",
}
`;

exports[`examples > 003-logical.bas 1`] = `
{
  "exitCode": 0,
  "stderr": "",
  "stdout": "false
true
",
}
`;

exports[`examples > 004-if.bas 1`] = `
{
  "exitCode": 0,
  "stderr": "",
  "stdout": "1 + 1 is 2!
1 + 1 remains equal to 2!
still 2!
",
}
`;

exports[`examples > 005-if-else.bas 1`] = `
{
  "exitCode": 0,
  "stderr": "",
  "stdout": "else if
",
}
`;

exports[`examples > 006-open-block.bas 1`] = `
{
  "exitCode": 0,
  "stderr": "/home/josh/matanuska/examples/006-open-block.bas:40:4: error: else has not been closed
  40  else
      ^
",
  "stdout": "",
}
`;

exports[`examples > 007-for.bas 1`] = `
{
  "exitCode": 0,
  "stderr": "",
  "stdout": "1
2
3
4
5
6
7
8
9
10
",
}
`;

exports[`examples > 008-while.bas 1`] = `
{
  "exitCode": 0,
  "stderr": "",
  "stdout": "loop
",
}
`;

exports[`examples > 009-repeat.bas 1`] = `
{
  "exitCode": 0,
  "stderr": "",
  "stdout": "loop
",
}
`;
